### PR TITLE
[safetensors] sharded with file path

### PR DESCRIPTION
**Pull request by mishig25**
Fixes https://github.com/huggingface/huggingface.js/issues/588

cc: @madgetr

### CODE REVIEW

**Review by madgetr — COMMENTED** [review]
Thanks for tackling this, I was just about to comment confirming that it is indeed broken!

File: packages/hub/src/lib/parse-safetensors-metadata.spec.ts [hunk 1]
```diff
@@ -83,4 +83,25 @@ describe("parseSafetensorsMetadata", () => {
 		assert.deepStrictEqual(parse.parameterCount, { F32: 859_520_964 });
 		assert.deepStrictEqual(sum(Object.values(parse.parameterCount)), 859_520_964);
 	});
+
+	it("fetch info for sharded (with the default conventional filename) with file path", async () => {
+		const parse = await parseSafetensorsMetadata({
```
Comment (madgetr) [inline packages/hub/src/lib/parse-safetensors-metadata.spec.ts:88]: I noticed that the test is not commit aware, if this model is deleted then this test will fail.

Comment (julien-c) [reply packages/hub/src/lib/parse-safetensors-metadata.spec.ts:88]: yep we can pass a `revision` (and we'd recreate this test if the repo is ever completely deleted)

File: packages/hub/src/lib/parse-safetensors-metadata.ts [hunk 2]
```diff
@@ -139,12 +140,13 @@ async function parseShardedIndex(
 		throw new SafetensorParseError(`Failed to parse file ${path}: not a valid JSON.`);
 	}
 
+	const pathPrefix = path.match(RE_FILE_DIR)?.[1] ?? "";
```
Comment (julien-c) [inline packages/hub/src/lib/parse-safetensors-metadata.ts:143]: maybe clearer intent to do something like (untested code):
```ts
const parentFolder = path.substr(0, path.lastIndexOf("/"))
```
?

File: packages/hub/src/lib/parse-safetensors-metadata.spec.ts (hunk 1 above)
Comment (mishig25) [reply packages/hub/src/lib/parse-safetensors-metadata.spec.ts:92]: handled in 06ab3c0f33420ddad25974f26d3e15d4521106f6

File: packages/hub/src/lib/parse-safetensors-metadata.ts (hunk 2 above)
Comment (mishig25) [reply packages/hub/src/lib/parse-safetensors-metadata.ts]: handled in 75ce40a190cdc910d9ca667d0650d1d1860152bc

**Review by madgetr — APPROVED** [review]
+1

**Review by julien-c — APPROVED** [review]
lgtm 

(CI failure is unrelated i think?)

File: packages/hub/src/lib/parse-safetensors-metadata.ts [hunk 3]
```diff
@@ -139,12 +139,13 @@ async function parseShardedIndex(
 		throw new SafetensorParseError(`Failed to parse file ${path}: not a valid JSON.`);
 	}
 
+	const pathPrefix = path.substr(0, path.lastIndexOf("/") + 1);
```
Comment (coyotte508) [inline packages/hub/src/lib/parse-safetensors-metadata.ts:142]: `substr` is deprecated for strings: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr

(`.slice` is my goto usually)

Comment (mishig25) [reply packages/hub/src/lib/parse-safetensors-metadata.ts:142]: https://github.com/huggingface/huggingface.js/pull/598

Comment (julien-c) [reply packages/hub/src/lib/parse-safetensors-metadata.ts:142]: it's still in the LLMs though =)

**Pull request merged**